File: assignment 9/python/programs/5.py
def find_max_length_subarray_equal_01(arr):
    max_length = 0
    prefix_sum = 0
    sum_indices = {0: -1}

    for i, num in enumerate(arr):
        if num == 0:
            prefix_sum -= 1
        else:
            prefix_sum += 1

        if prefix_sum in sum_indices:
            subarray_length = i - sum_indices[prefix_sum]
            if subarray_length > max_length:
                max_length = subarray_length
        else:
            sum_indices[prefix_sum] = i

    return max_length


# Example usage
arr = [0, 1, 0, 1, 1, 1, 0, 0]
max_length = find_max_length_subarray_equal_01(arr)
print("Maximum length subarray with an equal number of 0's and 1's:", max_length)

File: assignment 9/c/programs/5.c
#include <stdio.h>
#include <stdlib.h>

void findMaxSubarrayWithEqualZerosAndOnes(int arr[], int size)
{
    int maxLength = 0;
    int currentCount = 0;
    int endIndex = -1;
    int startIndex = -1;

    int *countDifferences = (int *)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++)
    {
        if (arr[i] == 0)
        {
            currentCount--;
        }
        else
        {
            currentCount++;
        }
        countDifferences[i] = currentCount;
    }

    // Create a hash table to store the first occurrence of each count difference
    int hashTable[2 * size + 1];
    for (int i = 0; i < 2 * size + 1; i++)
    {
        hashTable[i] = -1;
    }

    for (int i = 0; i < size; i++)
    {
        if (countDifferences[i] == 0)
        {
            maxLength = i + 1;
            endIndex = i;
        }
        else if (hashTable[countDifferences[i] + size] == -1)
        {
            hashTable[countDifferences[i] + size] = i;
        }
        else
        {
            if (i - hashTable[countDifferences[i] + size] > maxLength)
            {
                maxLength = i - hashTable[countDifferences[i] + size];
                startIndex = hashTable[countDifferences[i] + size] + 1;
                endIndex = i;
            }
        }
    }

    if (maxLength > 0)
    {
        printf("Maximum length subarray with equal 0s and 1s is [%d, %d]\n", startIndex, endIndex);
    }
    else
    {
        printf("No subarray with equal 0s and 1s found.\n");
    }

    free(countDifferences);
}

int main()
{
    int arr[] = {0, 1, 0, 1, 1, 0};
    int size = sizeof(arr) / sizeof(arr[0]);

    findMaxSubarrayWithEqualZerosAndOnes(arr, size);

    return 0;
}
